nucleus/common/glassfish-api/src/main/java/org/glassfish/api/invocation/InvocationManagerImpl.java

```diff
@@ -68,347 +71,334 @@
-            }
-        }
+    private InvocationManagerImpl(Iterable<ComponentInvocationHandler> handlers) {
+        this.allTypesHandler = initInvocationHandlers(handlers);
 
-        frames = new InheritableThreadLocal<InvocationArray<ComponentInvocation>>() {
+        framesByThread = new InheritableThreadLocal<InvocationFrames>() {
 
-            protected InvocationArray<ComponentInvocation> initialValue() {
-                return new InvocationArray<>();
+            @Override
+            protected InvocationFrames initialValue() {
+                return new InvocationFrames();
             }
 
-            protected InvocationArray<ComponentInvocation> childValue(InvocationArray<ComponentInvocation> parentValue) {
+            @Override
+            protected InvocationFrames childValue(InvocationFrames parentValue) {
                 return computeChildTheadInvocation(parentValue);
             }
         };
     }
 
-    @SuppressWarnings("unchecked")
+    private static ComponentInvocationHandler initInvocationHandlers(Iterable<ComponentInvocationHandler> handlers) {
+        if (handlers == null) {
+            return null;
+        }
+        List<ComponentInvocationHandler> hs = new ArrayList<>();
+        handlers.forEach(hs::add);
+        if (hs.isEmpty()) {
+            return null;
+        }
+        if (hs.size() == 1) {
+            return hs.get(0);
+        }
+        return new ListComponentInvocationHandler(hs);
+    }
+
     @Override
     public void setThreadInheritableInvocation(List<? extends ComponentInvocation> parentValue) {
-        frames.set(computeChildTheadInvocation((InvocationArray<ComponentInvocation>) parentValue));
+        framesByThread.set(computeChildTheadInvocation(InvocationFrames.valueOf(parentValue)));
     }
 
+    @Override
     public <T extends ComponentInvocation> void preInvoke(T invocation) throws InvocationException {
 
-        InvocationArray<ComponentInvocation> invocationArray = frames.get();
+        InvocationFrames frames = framesByThread.get();
         if (invocation.getInvocationType() == SERVICE_STARTUP) {
-            invocationArray.setInvocationAttribute(SERVICE_STARTUP);
+            frames.setState(SERVICE_STARTUP);
             return;
         }
 
-        int beforeSize = invocationArray.size();
-        ComponentInvocation previousInvocation = beforeSize > 0 ? invocationArray.get(beforeSize - 1) : null;
-
-        // If ejb call EJBSecurityManager, for servlet call RealmAdapter
-        ComponentInvocationType invocationType = invocation.getInvocationType();
+        ComponentInvocation prev = frames.peekLast();
+        ComponentInvocationType type = invocation.getInvocationType();
+        ComponentInvocationHandler typeHandler = typeHandlers.get(type);
 
-        if (invHandlers != null) {
-            for (ComponentInvocationHandler handler : invHandlers) {
-                handler.beforePreInvoke(invocationType, previousInvocation, invocation);
+        try {
```

Comment:
NB. Missing `try-finally` here to make sure the `frames.addLast(invocation);` will definitely happen is my best guess on what might have caused the issue observed,